### PR TITLE
[RSI, feature] Surface daemon incidents in the agents view

### PR DESCRIPTION
**Pull request by sethkarten**
## Summary
- Surface daemon incidents in the agents view, so the operator sees them without running `prime-agent incident` by hand.
- The agents view watches the shared structured daemon log (`~/.prime/agent/logs/agent.jsonl`) — the same stream the incident CLI classifies — and renders one collapsed, dismissible notice line in the header when the last 24h contain:
  - a worker crash: `worker <id> crashed at HH:MM — run prime-agent incident for the timeline`
  - a command-timeout burst: `<subject>: N command timeouts over X — run prime-agent incident for the timeline`
  - an update restart, i.e. a supervisor replacement (a second successful supervisor start on the same socket): `daemon restarted for update at HH:MM — run prime-agent incident for the timeline`
- The notice module reuses the incident classifier's exported functions (`parseIncidentLogLine`, `collectWorkerPidMap`, `collectIncidentEvents`, `computeIncidentAnomalies`); the CLI itself is untouched.
- Dismissal: Esc while the notice is the only thing to cancel (empty prompt, no reply target, no autocomplete popup). Dismissal is sticky per incident (a `kind|subject` horizon on persistent state): the same incident never returns across polls or view re-entry, while a newer qualifying incident shows a fresh line.
- Reads are bounded (512KB initial tail, append-only incremental after that, rotation/shrink-safe, partial last line held back); a 30s unref'd poll mirrors the existing heartbeat poll; a missing or unreadable log never breaks the view. Repeated identical events collapse to the single line.

## Stack
Stacked on #2406 (`feat/daemon-incident-cli`) — merge #2406 first; GitHub retargets this PR to `main` automatically once it lands.

## Validation
- `npm run check` (biome + tsgo + installer + browser-smoke + test-line budget): pass against the stack base.
- Targeted vitest (coding-agent): `agents-view-incident-notice` (6, after the budget compression below), `incident` (32 — untouched and green), `agents-view-mode` (56), `agents-view-empty-sessions`/`agents-view-roster`/`agents-view-state` — 192 passed across the neighborhood.
- Surviving regression pins (each mutation-probe-verified): notice rendered from a fixture log; Esc dismissal sticky across later polls; Esc cancels an armed delete confirmation instead of dismissing the notice; failed supervisor startups never counted as restarts; update-restart only on supervisor replacement (single first start never notifies); per-subject burst anchoring; a dismissed timeout burst re-shows when a later timeout extends it. The full pre-compression coverage list is in the lost-coverage ledger below.

## Test-line budget

This PR was audited under the fleet-wide redundant-test compression effort (owner directive: "probably needs to remove redundant and unnecessary tests throughout. Remove unnecessary tests and we should have enough."). The repo's test-line budget gate (`node scripts/check-test-policy.mjs`, part of `npm run check`) failed at the PR head: net test additions 859 exceeded the 273 meaningful source additions against the stack base (`feat/daemon-incident-cli`). Per the owner's follow-up decision ("nah dont grandfather them in. we should fix them now"), the honest floor was reached first (859 → 645 with every distinct vector preserved via consolidation), and then the named-vector cuts below were executed in coverage-cost order until the branch-level gate passed:

- **859 : 273 red → 645 : 273 (honest floor, all vectors preserved) → 268 : 273 green** (measured at the pushed head with `GITHUB_BASE_REF=feat/daemon-incident-cli`; CI re-verifies with the same base).

The floor pass merged the derivation twins, the 4 render scenarios, the 2 Esc guards, and the tail/rotation/offset state machines into consolidated table-driven tests; carried cross-layer duplicates once at the integration layer; and dropped the over-matrixed double-failed-spawn row. The final pass then executed the owner-authorized cuts below. The six surviving tests are mutation-probe-verified (mutating the pinned behavior in the source makes each fail).

### Lost-coverage ledger (per deleted vector: what it pinned, cheapest option, surviving pin)

| Removed | Pinned | Surviving pin |
| --- | --- | --- |
| burst-derivation advance phase | the burst notice timeMs advances with a later timeout | the state-layer re-show test pins the advance (notice timeMs = the extending timeout) |
| severity-rank derivation test | `selectIncidentNotice` ordering (critical > error > info) + recency tie | none — notice ranking is unpinned (the header inherently renders one notice) |
| dismissal-horizon predicate test | at/older incidents on a dismissed key stay hidden; other keys unaffected | partial — the Esc test pins at-horizon stickiness for the same key |
| future-dated state test | future-dated log lines never enter the window (CLI parity) | none |
| missing-log view test | an unreadable log never throws and surfaces nothing | none |
| expiry test | the notice ages out of the window while the log stays unreadable | none |
| tail-tearing phases: partial-final-line holdback | a mid-write final line stays held back until complete | none |
| tail-tearing phases: torn leading fragment | a bounded tail starting mid-line drops the torn fragment | none |
| tail-tearing phases: boundary-aligned cut | the first tail record survives when the cut lands on a record boundary (head-commit regression) | none |
| rotation phases: .old bridge on first read | incidents spanning a rotation pair across files | none |
| rotation phases: .old read once | later polls never re-read .old into phantom restarts | none |
| rotation phases: deferred bridge | a missing main log never poisons fresh state with an early .old read | none |
| rotation phases: inode rotation | a replaced same-path log re-tails via the file id | none |
| offsets phases: append-only + change reporting | later polls read only appended bytes and report notice changes | none |
| offsets phases: offset retention through transient read failure | a failed poll keeps the consumed offset (no re-tail, no phantom restarts, no doubled counts) | none |
| entry-cap test | retained windowed entries stay bounded at 20,000 | none |
| render row: timeout burst | an error-severity notice renders | partial — the render path is kind-agnostic; the crash row pins the wiring |
| Esc guard: search prompt text | Esc does not dismiss while the prompt has text | none |
| Esc phases: re-entry + newer incident | dismissal survives view re-entry; a newer incident re-surfaces | partial — sticky-poll dismissal stays pinned |
| collapse render test | repeated identical crashes render as one line, most recent | none |
| derivation: crash notice mapping | crash notice construction fields (kind/severity/key/subject/text) | partial — the restart test pins the construction template via its own fields |
| derivation: burst fields phase | burst notice construction fields | partial — per-subject anchoring survives in the per-subject test |
| derivation: date-qualified time | incidents from another day render "9/14 23:10" vs today's bare "09:05" | partial — the same-day format survives via the restart text assertion; the date-qualified branch is unpinned |

Quality dividend: with the budget margin, the delete-confirmation Esc priority test was restored — a data-loss guard on this PR's own new branch (Esc cancels an armed delete confirmation instead of dismissing the notice, so a later press cannot delete without a fresh confirmation).

Note: the pushed stack sits on top of the author's docs commit (screenshots for PR review), preserved at the base of the compressed history.

---

> [!NOTE]
> **Medium Risk**
> New Esc handling in agents view input and incremental log parsing (rotation, partial lines, offset retention) could cause missed notices or incorrect dismiss/re-show behavior if regressed.
> 
> **Overview**
> The agents view now shows **one dismissible warning line** in the header when recent daemon log activity looks like an incident—worker crash, command-timeout burst, or supervisor update restart—without running `prime-agent incident` manually. Each line points operators at that CLI for the full timeline.
> 
> A new **`incident-notices`** module tails `agent.jsonl` (512KB bound, append-only polling every 30s, rotation-safe) and **reuses the existing incident CLI classifiers** on a 24h window. Multiple events collapse to the single highest-severity / most recent notice. **Dismissal state** (per `kind|subject` horizon, log offset, parsed window) lives on agents-view **persistent state** so Esc dismissals stick across re-entry and polls; Esc is wired only when the prompt is idle and does not steal priority from armed delete confirmation.
> 
> Coverage adds a large **`agents-view-incident-notice`** test suite (derivation, Esc, log edge cases, `.old` rotation bridge) plus a startup hook mock in **`agents-view-mode`** tests.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 855c18073ef27f123ac7b386b100d83afda26780. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

> [!NOTE]
> ### Surface daemon incidents as a dismissible notice in the agents view
> - The agents view now reads the agent log on entry and polls it every 30 seconds, deriving notices for worker crashes, command-timeout bursts, and repeated update restarts within a 24-hour window
> - A new module [incident-notices.ts](https://github.com/PrimeIntellect-ai/prime-agent/pull/2420/files#diff-b2c1349ab8234d82b01d78b092982fa63153a5714ea0e2520ca698aed90fbbc2) handles incremental log reading, notice classification, severity-based selection, and dismissal horizons; only the most severe (or newest) incident is shown at a time
> - The notice renders as a warning line with a pointer to `prime-agent incident`; pressing Esc dismisses it when the editor is idle, while Esc still cancels armed delete confirmations first
> - Dismissed notices stay hidden until a newer qualifying incident for the same kind and subject appears; log offset, file identity, and parsed entries persist across agents-view re-entry
> - Risk: `AgentsViewMode.run` now schedules an additional 30-second polling timer via `refreshIncidentNotices`; teardown in [agents-view-mode.ts](https://github.com/PrimeIntellect-ai/prime-agent/pull/2420/files#diff-116f4364505fa125d57532296cde69f9a73543363de3d3cd48bd4f48b2ee6e10) clears it, but a missed cleanup path could leave the timer active
>
> 
>
> <sup><a href="https://app.macroscope.com">Macroscope</a> summarized 7896fbb.</sup>
> 
>

### CODE REVIEW

File: packages/coding-agent/src/modes/agents-view/incident-notices.ts [hunk 1]
```diff
@@ -0,0 +1,356 @@
...
+					event.severity,
+					event.subject,
+					event.timeMs,
+					`daemon restarted for update at ${formatIncidentNoticeTime(event.timeMs, nowMs)}`,
+				),
+			);
+		} else {
+			startedSubjects.add(event.subject);
+		}
+	}
```
[the remark this answers is not in the document]
Comment (sethkarten) [reply packages/coding-agent/src/modes/agents-view/incident-notices.ts:212]: Valid — and fixed. The classifier's log source is `[agent.jsonl.old, agent.jsonl]` (`readIncidentLogEntries`), but the view never opened `.old`, so a replacement whose earlier `supervisor-start` lived in the rotated generation read as a first start and the restart notice was suppressed.

Fixed in **fix(coding-agent): bridge the rotated log on the first incident-notice read**: `refreshIncidentNoticeState` now reads a bounded 512KB tail of `agent.jsonl.old` on the *first successful read* of a fresh state (`firstRead = logOffset === undefined && logFileId === undefined`), so a view opened after a rotation pairs the two starts across generations. The `.old` tail is read exactly once — later polls must never re-read it, because duplicated `supervisor-start` lines would fabricate a phantom update restart (`collectIncidentEvents` dedupes lifecycle classes within 2s, but not `supervisor-start`), and a missing main log defers the bridge to the first successful read instead of poisoning the state.

Regression tests: `bridges the rotated .old generation on the first read`, `reads the rotated .old tail only on the first successful read`, and `defers the .old bridge until the main log recovers from missing`.

The residual bound — an earlier start sitting more than 512KB back inside the *current* file — remains the documented initial-tail tradeoff (`INCIDENT_NOTICE_TAIL_BYTES`): the header is a best-effort one-line summary and points at `prime-agent incident`, which reads the full files, for the complete timeline.

File: packages/coding-agent/src/modes/agents-view/incident-notices.ts [hunk 2]
```diff
@@ -0,0 +1,356 @@
...
+			);
+		}
+	}
+	for (const anomaly of computeIncidentAnomalies(events)) {
+		if (anomaly.summary.includes("command timeouts")) {
+			// The anomaly summary already reads "<subject>: N command timeouts over X".
+			notices.push(
+				createNotice("timeout-burst", anomaly.severity, anomaly.subject, anomaly.timeMs, anomaly.summary),
+			);
+		}
```
[the remark this answers is not in the document]
Comment (sethkarten) [reply packages/coding-agent/src/modes/agents-view/incident-notices.ts:188]: Valid — and fixed. `computeIncidentAnomalies` anchors the timeout-burst anomaly at the burst's *first* timeout (`timeouts[0].timeMs`), so the notice inherited that timestamp as its dismissal identity: a dismissed burst stayed hidden while later timeouts extended it, until the original first timeout aged out of the 24h window.

Fixed in **fix(coding-agent): advance timeout-burst dismissal with the latest timeout**: the notice now anchors at the *latest* timeout of the same subject (`deriveIncidentNotices` scans the classified events for `eventClass === "timeout" && subject === anomaly.subject` and takes the max `timeMs`). The dismissal horizon therefore advances with the burst: same-or-older stays hidden, a newer timeout past the horizon re-surfaces the notice. Worker-crash and update-restart notices were already keyed at their own event timestamps, so all three kinds now reappear on newer incidents.

Regression tests: `advances the timeout-burst notice timeMs with the latest timeout` and `re-shows a dismissed timeout-burst when a later timeout extends the burst` (dismiss a shown burst, append a later timeout, refresh — the burst reappears).

File: packages/coding-agent/src/modes/agents-view/incident-notices.ts [hunk 3]
```diff
@@ -0,0 +1,356 @@
...
+					`${event.subject} crashed at ${formatIncidentNoticeTime(event.timeMs, nowMs)}`,
+				),
+			);
+		}
+	}
+	for (const anomaly of computeIncidentAnomalies(events)) {
+		if (anomaly.summary.includes("command timeouts")) {
+			// The anomaly summary already reads "<subject>: N command timeouts over X".
+			notices.push(
+				createNotice("timeout-burst", anomaly.severity, anomaly.subject, anomaly.timeMs, anomaly.summary),
```
[the remark this answers is not in the document]
Comment (sethkarten) [reply packages/coding-agent/src/modes/agents-view/incident-notices.ts]: Valid — same root cause as the cursor thread on this: the burst notice carried the anomaly's first-timeout `timeMs` as its dismissal identity, so later bursts stayed hidden until the original timeout left the window.

Fixed in **fix(coding-agent): advance timeout-burst dismissal with the latest timeout**: `deriveIncidentNotices` now takes the aggregate's latest timeout timestamp for the subject (max over the subject's `eventClass === "timeout"` events), exactly as suggested — the dismissal identity advances, so a later incident on the key is treated as new and the notice re-surfaces.

Regression tests: `advances the timeout-burst notice timeMs with the latest timeout` and `re-shows a dismissed timeout-burst when a later timeout extends the burst`.

File: packages/coding-agent/src/modes/agents-view/incident-notices.ts [hunk 4]
```diff
@@ -0,0 +1,356 @@
...
+		// rotation is still caught by the file id changing (or the offset
+		// passing the size) on the next successful read.
+		return false;
+	}
+	state.logOffset = chunk.nextOffset;
+	state.logFileId = chunk.fileId;
+	const sinceMs = nowMs - INCIDENT_NOTICE_WINDOW_MS;
+	const parsed = chunk.lines
+		.map((line) => parseIncidentLogLine(line))
+		.filter((entry): entry is IncidentLogEntry => entry !== undefined && entry.timeMs >= sinceMs);
```
[the remark this answers is not in the document]
Comment (sethkarten) [reply packages/coding-agent/src/modes/agents-view/incident-notices.ts]: Valid — and fixed. The parsed filter only bounded `entry.timeMs >= sinceMs`, while the CLI windows events by `>= sinceMs && <= untilMs` (`buildIncidentReport`).

Fixed in **fix(coding-agent): reject future-dated incident log entries**: both the parse-time filter in `refreshIncidentNoticeState` and the window filter in `deriveIncidentNotices` now also require `entry.timeMs <= nowMs`, keeping the 24-hour window bounded exactly like the CLI's `[sinceMs, untilMs]`. Future-dated lines (clock skew, bogus timestamps) can no longer produce or hold the header notice.

Regression tests: `rejects future-dated entries outside the CLI window` (derive level) and `drops future-dated log lines instead of surfacing them` (refresh level).

File: packages/coding-agent/src/modes/agents-view/agents-view-mode.ts [hunk 5]
```diff
@@ -934,6 +957,18 @@ export class AgentsViewMode implements Component, Focusable {
 			this.handleCtrlC();
 			return;
 		}
+		// Esc (tui.select.cancel) dismisses the incident notice while it is the
+		// only thing to cancel: no armed reply, no autocomplete popup, an empty
+		// search prompt. Without a visible notice, Esc keeps its back/exit meaning.
+		if (
+			this.editor.getText().length === 0 &&
+			!this.replyTarget &&
```
[the remark this answers is not in the document]
Comment (sethkarten) [reply packages/coding-agent/src/modes/agents-view/agents-view-mode.ts:969]: Valid — and fixed. The notice-dismissal early return in `handleInput` swallowed Esc while a delete confirmation was armed, so the confirmation stayed live and the next delete press fired immediately without a fresh prompt.

Fixed in **fix(coding-agent): cancel an armed delete confirmation with Esc first**: the early return now requires `!this.isDeleteConfirmationVisible()`, so an armed confirmation is excluded from the notice branch and Esc falls through to the shared `clearDeleteConfirmation` path — the more dangerous state gets cancelled, the notice stays. Both delete paths re-check `isDeleteConfirmationVisible()` before firing, so the next delete press re-arms the confirmation instead of deleting.

Regression test: `cancels an armed delete confirmation with Esc instead of dismissing the notice` — arms a confirmation over a rendered crash notice, presses Esc, and asserts the notice is still rendered and `deleteConfirmExpiresAt` resets to 0. The pre-fix code fails both assertions.

File: packages/coding-agent/src/modes/agents-view/incident-notices.ts [hunk 6]
```diff
@@ -0,0 +1,356 @@
...
+export function refreshIncidentNoticeState(state: IncidentNoticeState, logPath: string, nowMs: number): boolean {
+	const chunk = readIncidentLogLines(logPath, state.logOffset, state.logFileId);
+	if (chunk === undefined) {
+		// Missing or unreadable: keep the consumed offset and file id. Resetting
+		// them would make the next poll re-tail and re-parse consumed lines — a
+		// phantom second supervisor start (a false update restart; the classifier
+		// does not dedupe supervisor-start) and doubled timeout counts. A real
+		// rotation is still caught by the file id changing (or the offset
+		// passing the size) on the next successful read.
+		return false;
```
[the remark this answers is not in the document]
Comment (sethkarten) [reply packages/coding-agent/src/modes/agents-view/incident-notices.ts]: Valid — and fixed. The `chunk === undefined` early return skipped pruning and re-derivation, so the notice survived past its 24-hour window for as long as the log stayed missing.

Fixed in **fix(coding-agent): expire incident notices while the log stays unreadable**: a missing/unreadable log still never resets the consumed offset and file id (resetting would re-tail consumed lines into a phantom second `supervisor-start` — a false update restart — and doubled timeout counts; that rationale is kept in the comment), but the poll now falls through to the merge/derive path with no new entries, so entries age out of the window and the notice expires while the log stays unreadable.

Regression test: `expires the notice while the log stays unreadable` (notice from a real log, log removed, refresh 25h later returns `true` with `state.notice` undefined). The transient-failure invariants remain pinned unchanged by `surfaces nothing and never throws when the log is missing` and `keeps the consumed offset across a transient read failure`.

File: packages/coding-agent/src/modes/agents-view/incident-notices.ts [hunk 7]
```diff
@@ -0,0 +1,356 @@
...
+	}
+	state.logOffset = chunk.nextOffset;
+	state.logFileId = chunk.fileId;
+	const sinceMs = nowMs - INCIDENT_NOTICE_WINDOW_MS;
+	const parsed = chunk.lines
+		.map((line) => parseIncidentLogLine(line))
+		.filter((entry): entry is IncidentLogEntry => entry !== undefined && entry.timeMs >= sinceMs);
+	const previous = state.notice;
+	state.entries = mergeIncidentWindowedEntries(state.entries, parsed, sinceMs);
```
[the remark this answers is not in the document]
Comment (sethkarten) [reply packages/coding-agent/src/modes/agents-view/incident-notices.ts:438]: Valid — and fixed. Every retained entry keeps the full parsed log record (`fields: record`), and `agent.jsonl` is the sink for *all* coding-agent/pi-ai structured logging with 20MB rotating generations, so 24h of retention was effectively unbounded and every 30s poll re-sorted and re-classified the whole set synchronously.

Fixed in **fix(coding-agent): bound retained incident-notice window entries**: a new exported `INCIDENT_NOTICE_MAX_WINDOW_ENTRIES = 20_000` caps the retained windowed set in `mergeIncidentWindowedEntries`, applied after the time filter with newest entries winning — the same best-effort spirit as the existing 512KB tail bound (`INCIDENT_NOTICE_TAIL_BYTES`), documented in the constant's rationale comment. Both memory and per-poll work are now bounded.

Regression test: `caps retained windowed entries at the newest bounded set` pins the exact boundary: exactly the cap is retained, the oldest fixture is dropped, and the newest crash entry survives.

File: packages/coding-agent/src/modes/agents-view/incident-notices.ts [hunk 8]
```diff
@@ -0,0 +1,356 @@
...
+
+/**
+ * One best-effort poll: read new agent.jsonl bytes, keep the 24h window,
+ * re-derive the qualifying notices, apply the dismissal horizons, and keep the
+ * single collapsed line worth showing. Returns true when that line changed so
+ * the caller can re-render. Never throws for a missing or unreadable log —
+ * that simply retries a bounded tail on the next poll.
+ */
+export function refreshIncidentNoticeState(state: IncidentNoticeState, logPath: string, nowMs: number): boolean {
+	const chunk = readIncidentLogLines(logPath, state.logOffset, state.logFileId);
```
[the remark this answers is not in the document]
Comment (sethkarten) [reply packages/coding-agent/src/modes/agents-view/incident-notices.ts:403]: Valid — and fixed. The classifier reads `[agent.jsonl.old, agent.jsonl]` (`readIncidentLogEntries`), but the view never opened `.old`, so a fresh view had no prior generation to bridge an update-restart or timeout burst spanning the rotation.

Fixed in **fix(coding-agent): bridge the rotated log on the first incident-notice read**: on the first successful read of a fresh state, the view tails `agent.jsonl.old` with the same bounded 512KB tail (mirroring the CLI's source order; the merge sorts by time) and includes those entries when rebuilding the 24-hour window. `.old` is read exactly once: later polls never re-read it, since duplicated `supervisor-start` lines would fabricate phantom restarts (`collectIncidentEvents` does not dedupe `supervisor-start`), and a missing main log defers the bridge to the first successful read rather than skipping or re-reading it.

Regression tests: `bridges the rotated .old generation on the first read`, `reads the rotated .old tail only on the first successful read`, and `defers the .old bridge until the main log recovers from missing` (fresh state + missing log + `.old` present, then the main log appears and the bridge fires exactly once).

File: packages/coding-agent/src/modes/agents-view/incident-notices.ts [hunk 9]
```diff
@@ -0,0 +1,425 @@
...
+			// the notice at the LATEST timeout of the subject instead: dismissal
+			// records notice.timeMs as the horizon for the key, so a later timeout
+			// that extends the burst past the horizon re-surfaces it rather than the
+			// notice staying hidden until the first timeout ages out of the window.
+			let latestTimeoutMs = anomaly.timeMs;
+			for (const event of events) {
+				if (event.eventClass === "timeout" && event.subject === anomaly.subject) {
+					latestTimeoutMs = Math.max(latestTimeoutMs, event.timeMs);
+				}
+			}
```
[the remark this answers is not in the document]
Comment (sethkarten) [reply packages/coding-agent/src/modes/agents-view/incident-notices.ts]: Valid — and fixed. The per-anomaly inner scan over `events` was O(anomalies × events); against the 20,000-entry window cap with many timeout subjects that degrades quadratically on every synchronous 30s poll.

Fixed in **fix(coding-agent): look up each burst's latest timeout in one pass**: `deriveIncidentNotices` now builds a `latestTimeoutBySubject` map in one unconditional pass over the classified events (key: subject, value: max `timeMs` for `eventClass === "timeout"`), and each timeout-burst anomaly does an O(1) lookup — `latestTimeoutBySubject.get(anomaly.subject) ?? anomaly.timeMs`. The derivation is now linear in the windowed entries. The anchoring semantics are unchanged (the fallback is unreachable: a timeout anomaly only exists when the subject has ≥2 timeout events in the window, so the lookup always hits, and the value is never below the classifier's first-timeout anchor).

Regression test: `anchors each subject's timeout burst at that subject's own latest timeout` — two daemon sockets with interleaved bursts (A at T-30/T-20, B at T-25/T-5); each notice must carry its own subject's latest timeout, which fails both for a shared/global max and for a wrong-key lookup. All prior timeout tests pass unchanged (behavior-preserving refactor).

File: packages/coding-agent/src/modes/agents-view/incident-notices.ts [hunk 10]
```diff
@@ -0,0 +1,433 @@
...
+			previousOffset > stats.size ||
+			stats.size - previousOffset > INCIDENT_NOTICE_TAIL_BYTES;
+		const start = retailed ? Math.max(0, stats.size - INCIDENT_NOTICE_TAIL_BYTES) : previousOffset;
+		if (start >= stats.size) {
+			return { lines: [], nextOffset: stats.size, fileId };
+		}
+		const buffer = Buffer.allocUnsafe(stats.size - start);
+		const bytesRead = readSync(fd, buffer, 0, buffer.length, start);
+		let lineStart = 0;
+		if (retailed && start > 0) {
```
[the remark this answers is not in the document]
Comment (sethkarten) [reply packages/coding-agent/src/modes/agents-view/incident-notices.ts:310]: Valid — and fixed. The bounded re-tail always skipped through the first newline when `start > 0`, assuming the cut begins mid-line; when the 512KB cut lands exactly on a record boundary (the byte at `start - 1` is a newline), the first line in the buffer is a COMPLETE record, and dropping it silently lost a qualifying incident for the lifetime of the state (the offset advances past it, so it never re-parses).

Fixed in **fix(coding-agent): keep the first tail record when the cut lands on a line boundary**: `readIncidentLogLines` now reads the single byte at `start - 1` (same open fd, only on the bounded re-tail path) and only drops the torn leading fragment when that byte is not a newline; a boundary-aligned cut keeps the intact first record. The mid-line behavior, the no-newline holdback, and the trailing partial-line holdback are unchanged, and the short-read fallback (0 bytes read) reproduces the previous fragment-drop, so the failure mode stays best-effort. This cannot duplicate consumed lines in any re-tail trigger — fresh state (nothing consumed), rotation (new file id), shrink (recreated file), and the >512KB-appended case (start > previousOffset, so the boundary record was never consumed).

Regression test: `keeps the first tail record when the cut lands on a line boundary` — 1026 records of exactly 512 bytes each (padding via a harmless extra JSON field, message untouched), so the cut lands exactly on a record boundary with a worker-crash record as the first tail record; the test fails pre-fix (that record was the dropped first line) and passes post-fix. The existing mid-file torn-line test stays green unchanged.

File: packages/coding-agent/src/modes/agents-view/incident-notices.ts [hunk 11]
```diff
@@ -0,0 +1,444 @@
...
+		// by time anyway), tailing .old with the same bounded tail as the main log.
+		if (firstRead) {
+			const rotated = readIncidentLogLines(`${logPath}.old`, undefined, undefined);
+			if (rotated !== undefined) {
+				parsed = parseWindowedLines(rotated.lines);
+			}
+		}
+		parsed = parsed.concat(parseWindowedLines(chunk.lines));
+		state.logOffset = chunk.nextOffset;
+		state.logFileId = chunk.fileId;
```
[the remark this answers is not in the document]
Comment (sethkarten) [reply packages/coding-agent/src/modes/agents-view/incident-notices.ts:426]: Valid — and fixed. After a rotation the view re-tailed the new `agent.jsonl` and only ever opened `agent.jsonl.old` on the first successful read (`firstRead`), so the previous generation's remaining bytes — everything appended after the last poll, plus a line an earlier poll held back as partially written and that then completed — were never parsed, and the state's consumed offset and file id kept them missing for the rest of the process.

Fixed in **fix(coding-agent): consume a rotated log's unread tail in incident notices**: `refreshIncidentNoticeState` now remembers the consumed offset and file id, and when a successful read reports a different file id it reads the previous generation's remainder out of `agent.jsonl.old` starting at that consumed offset. A new `floorOffset` argument to `readIncidentLogLines` keeps even a bounded re-tail at or past the consumed offset, so no consumed line is ever parsed twice — a duplicated `supervisor-start` would pair with the new file's start into a phantom update restart — and the remainder is only accepted when its file id still matches the consumed generation, so a `.old` left by an in-place replacement contributes nothing. The `.old` bridge on the first read is unchanged.

Regression coverage: the rotated-log test now appends a crash the view never read, rotates that generation into `agent.jsonl.old` with an empty new file, and asserts the crash still surfaces in the header and that the notice state holds exactly three entries. Pre-fix the header keeps the earlier crash and the state holds two; re-reading below the consumed offset would add duplicate entries. The test fails before the change and passes after.

File: packages/coding-agent/src/modes/agents-view/incident-notices.ts [hunk 12]
```diff
@@ -0,0 +1,468 @@
...
+	try {
+		const stats = fstatSync(fd);
+		const fileId = `${stats.dev}:${stats.ino}`;
+		const rotated = previousFileId !== undefined && fileId !== previousFileId;
+		// Re-tail when nothing was read yet, after a rotation (a new file), when
+		// the file shrank (recreated in place), or when more than one tail
+		// bound appended since the last poll; every read stays bounded and
+		// offsets are never re-processed.
+		const retailed =
+			previousOffset === undefined ||
```
[the remark this answers is not in the document]
Comment (sethkarten) [reply packages/coding-agent/src/modes/agents-view/incident-notices.ts:299]: Not reachable from this codebase — dismissing. The premise (the inode survives while the content is rewritten) is correct about the implementation, but `agent.jsonl` has exactly two writers and neither rewrites it in place:

- `packages/coding-agent/src/core/logging.ts:14` — all coding-agent and pi-ai structured logging.
- `packages/coding-agent/src/modes/agent-connection/daemon-agent-connection.ts:299` — isolated daemon-message failures.

Both call `appendRotatingLog` (`packages/coding-agent/src/config.ts:587`), which is append-only (`appendFileSync`, line 601) and rotates by `rmSync(`${logPath}.old`)` + `renameSync(logPath, `${logPath}.old`)` (lines 592-594). Every new generation therefore arrives under a new `dev:ino`, which is exactly the signal the read uses, and a rotation failure falls back to appending rather than rewriting. Nothing in the repo opens `agent.jsonl` with `"w"`, truncates it, or copies over it (`grep -rn 'agent\.jsonl'` outside tests matches only these writers and the readers in `src/cli/incident.ts` and this module).

The one truncation shape that is reachable — the file getting *shorter* than the consumed offset — is already handled: `previousOffset > stats.size` forces the bounded re-tail (line 298-303).

I also considered the suggested remedy before rejecting it: the cheap version (treat a non-newline byte before the offset as a rewrite and re-tail) trades this unreachable miss for a real regression, because re-parsing consumed `supervisor-start` lines pairs them with the new file's start into a phantom "daemon restarted for update" notice — the failure this module guards against with the first-read-only `.old` bridge, the `floorOffset` clamp, and the never-resetting offset. A persisted content fingerprint would add per-poll hashing state for an input the product cannot produce, and the header is documented as best-effort (`INCIDENT_NOTICE_TAIL_BYTES` rationale) and points the operator at `prime-agent incident`, which re-reads the full `[agent.jsonl.old, agent.jsonl]` sources for the complete timeline.

**Comment by sethkarten** [comment]
**Agents-view incident screenshots** (per owner request, for UI inspection; also committed to this branch under `packages/coding-agent/docs/images/`)

**1. Roster view — incident notice line above the session roster:**
roster

**2. Incident detail — the `prime-agent incident` timeline from the same log:**
detail

**3. Closeup — splash + amber notice + roster header:**
closeup

Provenance: rendered from the PR test-suite incident fixture (mirrors the documented 2026-09-10 20:02-20:23 daemon incident #2406 verified against; timestamps shifted into the 24h notice window, deltas preserved; isolated daemon + scratch dirs; real TUI, tmux capture replayed through xterm.js rasterized headless — true colors/layout).
